### PR TITLE
[Feature][Engine] Add configurable pipeline concurrency control

### PR DESCRIPTION
**Pull request by lihjChina**
Description
This PR introduces a configurable pipeline concurrency control mechanism that allows users to control how many pipelines execute simultaneously within a job.

What is the purpose of the change
Currently, SeaTunnel executes all pipelines concurrently when a job starts. This can cause issues when:

Multiple pipelines compete for limited database connections
Sequential execution is required for data consistency
Resource constraints require limiting concurrent operations
This change adds a pipeline_concurrency configuration option to control the maximum number of pipelines that can run simultaneously.

Motivation
Problem:

Users cannot control pipeline execution order
All pipelines start simultaneously, causing resource contention
No way to enforce serial execution when needed
Use Case:

A job with 4 pipelines reading from the same database
Database has connection limits
Need to execute pipelines serially to avoid connection exhaustion
Implementation
Changes Made
1. JobConfig Enhancement
File: 
seatunnel-engine-common/src/main/java/org/apache/seatunnel/engine/common/config/JobConfig.java

Added pipelineConcurrency field (default: Integer.MAX_VALUE for unlimited)
Updated serialization methods (
writeData
, 
readData
)
2. Configuration Parsing
File: 
seatunnel-engine-core/src/main/java/org/apache/seatunnel/engine/core/parse/MultipleTableJobConfigParser.java

Added logic to parse pipeline_concurrency from env configuration
Validates and sets the value in 
JobConfig
3. Execution Control
File: 
seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java

Added concurrency control fields:
pipelineConcurrency: configured limit
runningPipelineCount: current running count
nextPipelineIndex: next pipeline to start
Implemented 
startNextPipelines()
 method for controlled startup
Modified 
stateProcess()
 to use controlled startup
Enhanced 
addPipelineEndCallback()
 to start next pipeline when one completes
Added detailed logging for monitoring execution
Configuration Format
env {
  job.mode = "BATCH"
  job.name = "example_job"
  pipeline_concurrency = 1  # 1 for serial, N for limited concurrency, omit for unlimited
}
Testing
Test Environment
SeaTunnel version: [your version]
Test scenario: 4 pipelines reading from MySQL
Test Cases
Serial Execution (pipeline_concurrency = 1)

### CODE REVIEW

**Comment by DanielCarter-stack** [comment]
- **PhysicalPlan.java:183-185** - `finishedPipelineNum.incrementAndGet()` is called twice for each finished pipeline (also at line 200), causing the condition `finishedPipelineNum == pipelineList.size()` to never become true. Jobs will hang in RUNNING state forever, causing resource leaks.

- **JobConfig.java:52-64** - Adding `pipelineConcurrency` to serialization without versioning breaks backward compatibility. Deserializing old savepoints will read `envOptions` data into `pipelineConcurrency`, causing data corruption and recovery failures.

- **MultipleTableJobConfigParser.java:351-363** - Invalid config values (≤0) only log warnings but don't set a default. While `Integer.MAX_VALUE` is the field default, users expecting concurrency control won't get it without clear error messages.

- **MultipleTableJobConfigParser.java:340-368** - Debug logs like `"Checking for pipeline_concurrency configuration..."` use INFO level, which will flood production logs on every job startup.

- **PhysicalPlan.java:92-94** - `runningPipelineCount` and `nextPipelineIndex` always initialize from 0, ignoring savepoint state. Recovery may exceed concurrency limits or restart completed pipelines, causing data duplication in non-idempotent sinks.

- **PipelineConcurrencyIT.java** - Tests only verify log strings, not actual concurrent behavior. Missing: boundary values (0, negative), savepoint recovery, failure scenarios, unit tests. This is how critical bugs like the double-increment slipped through.

- **EnvCommonOptions.java:114-121; JobConfig.java:37; PhysicalPlan.java:92-94** - Missing JavaDoc for new fields/methods (`pipelineConcurrency`, `runningPipelineCount`, `nextPipelineIndex`, `startNextPipelines()`), reducing maintainability and usability.

**Comment by lihjChina** [comment]
> * **PhysicalPlan.java:183-185** - `finishedPipelineNum.incrementAndGet()` is called twice for each finished pipeline (also at line 200), causing the condition `finishedPipelineNum == pipelineList.size()` to never become true. Jobs will hang in RUNNING state forever, causing resource leaks.
> * **JobConfig.java:52-64** - Adding `pipelineConcurrency` to serialization without versioning breaks backward compatibility. Deserializing old savepoints will read `envOptions` data into `pipelineConcurrency`, causing data corruption and recovery failures.
> * **MultipleTableJobConfigParser.java:351-363** - Invalid config values (≤0) only log warnings but don't set a default. While `Integer.MAX_VALUE` is the field default, users expecting concurrency control won't get it without clear error messages.
> * **MultipleTableJobConfigParser.java:340-368** - Debug logs like `"Checking for pipeline_concurrency configuration..."` use INFO level, which will flood production logs on every job startup.
> * **PhysicalPlan.java:92-94** - `runningPipelineCount` and `nextPipelineIndex` always initialize from 0, ignoring savepoint state. Recovery may exceed concurrency limits or restart completed pipelines, causing data duplication in non-idempotent sinks.
> * **PipelineConcurrencyIT.java** - Tests only verify log strings, not actual concurrent behavior. Missing: boundary values (0, negative), savepoint recovery, failure scenarios, unit tests. This is how critical bugs like the double-increment slipped through.
> * **EnvCommonOptions.java:114-121; JobConfig.java:37; PhysicalPlan.java:92-94** - Missing JavaDoc for new fields/methods (`pipelineConcurrency`, `runningPipelineCount`, `nextPipelineIndex`, `startNextPipelines()`), reducing maintainability and usability.

Thank you very much for your in-depth code review! The issues you identified are all crucial, especially the double increment bug, which, if not fixed, can lead to serious production problems.

File: seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/java/org/apache/seatunnel/engine/e2e/PipelineConcurrencyIT.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+        // Verify serial mode logging
+        Assertions.assertTrue(
+                logs.contains("Parsed pipeline_concurrency from config: 1"),
+                "Should parse pipeline_concurrency = 1");
+        Assertions.assertTrue(
+                logs.contains("Pipeline concurrency: 1"), "Should show pipeline concurrency as 1");
+        Assertions.assertTrue(logs.contains("[Serial Mode]"), "Should execute in serial mode");
+
+        log.info("✅ Serial pipeline execution test passed");
```
Comment (dybyte) [inline seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/java/org/apache/seatunnel/engine/e2e/PipelineConcurrencyIT.java:50]: It looks like the current code isn’t fully covered by tests. Would you mind adding some unit tests?

Comment (lihjChina) [reply seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/java/org/apache/seatunnel/engine/e2e/PipelineConcurrencyIT.java:50]: ok

**Comment by DanielLeens** [comment]
Thanks for keeping this work alive. I reviewed the current head (`475c3b48`) from scratch against the engine scheduling, restore/takeover, serialization, and test paths. I also read the earlier discussion, and I largely agree with the blockers already raised by @DanielCarter-stack — after tracing the current call chain locally, I still do not think this revision is safe to merge yet.

# What this PR fixes
- User pain: when one job contains multiple pipelines, SeaTunnel currently has no explicit way to cap how many pipelines are allowed to run at the same time.
- Fix approach: this PR adds `env.pipeline_concurrency`, threads it through `JobConfig`, and changes `PhysicalPlan` from “start every CREATED pipeline” to “start up to N pipelines”.
- One-line summary: the direction makes sense, but the current head changes engine scheduling and serialization boundaries without closing the correctness and compatibility gaps.

# Runtime chain I checked
~~~text
Normal startup
  -> JobMaster.run() [JobMaster.java:540-542]
  -> PhysicalPlan.startJob() [PhysicalPlan.java:337-341]
  -> stateProcess() [PhysicalPlan.java:349-364]
  -> startNextPipelines() [PhysicalPlan.java:406-417]

Restore / master-takeover path
  -> CoordinatorService submits a RESTORE job [CoordinatorService.java:311-321]
  -> each SubPlan.restorePipelineState() is called first [SubPlan.java:522-560]
  -> RUNNING pipelines continue from the restored state
  -> JobMaster.run() then calls PhysicalPlan.startJob() again
  -> startNextPipelines() still sees runningPipelineCount = 0 and nextPipelineIndex = 0
  -> serial / limited mode can start a later CREATED pipeline even though an earlier pipeline is already RUNNING

Completion accounting
  -> FINISHED callback in PhysicalPlan.addPipelineEndCallback() [PhysicalPlan.java:178-200]
  -> finishedPipelineNum is incremented once in the FINISHED branch
  -> and then incremented again unconditionally at line 200
~~~

# Findings
Issue 1: pipeline completion is counted twice, so the job-level end-state decision becomes incorrect
- Location: `seatunnel-engine/seatunnel-engine-server/src/main/java/org/apache/seatunnel/engine/server/dag/physical/PhysicalPlan.java:178-200`
- Why this matters: a single FINISHED pipeline increments `finishedPipelineNum` in the FINISHED branch and then increments it again immediately afterward at line 200.
- Risk: the job can satisfy `finishedPipelineNum == pipelineList.size()` earlier than it should, which can prematurely drive the job into its terminal-state branch.
- Better fix: increment the completed-pipeline counter exactly once per finished pipeline and reuse that value for both logging and end-state checks.
- Severity: High
- Related review history: `+1` to @DanielCarter-stack's earlier point here.

Issue 2: `JobConfig` changes the `IdentifiedDataSerializable` field order without any compatibility handling
- Location: `seatunnel-engine/seatunnel-engine-common/src/main/java/org/apache/seatunnel/engine/common/config/JobConfig.java:37,52-64`
- Why this matters: this is a real engine serialization boundary. The PR inserts `pipelineConcurrency` between `jobContext` and `envOptions`, so old readers and new writers no longer agree on the binary layout.
- Risk: rolling upgrades, takeover, and cross-version deserialization can misread the payload and break recovery.
- Better fix: add explicit version-aware serialization, or at minimum introduce a backward-compatible read path before changing the layout.
- Severity: High
- Related review history: `+1` to @DanielCarter-stack's earlier point here.

Issue 3: the restore / takeover path can bypass the new concurrency limit
- Location: `CoordinatorService.java:311-321`, `SubPlan.java:522-560`, `PhysicalPlan.java:91-94,337-363,406-417`
- Why this matters: on master switch, RUNNING pipelines are restored first, but `PhysicalPlan.startJob()` still restarts the job-level scheduling path with `runningPipelineCount = 0` and `nextPipelineIndex = 0`. In serial / limited mode, that means a later CREATED pipeline can be started even though an earlier pipeline is already RUNNING.
- Risk: the limit that is supposed to serialize or cap pipeline execution is violated exactly on the restore/takeover path, which is one of the most sensitive runtime paths in SeaTunnel engine.
- Better fix: rebuild the scheduler counters from the current pipeline states before calling `startNextPipelines()`, and do not consume `nextPipelineIndex` as if a skipped non-CREATED pipeline were a free slot.
- Severity: High
- Related review history: `+1` to @DanielCarter-stack's earlier restore-state concern, with the concrete call chain above.

Issue 4: the new E2E tests only assert log strings, so they do not prove that concurrency control actually works
- Location: `seatunnel-e2e/seatunnel-engine-e2e/connector-seatunnel-e2e-base/src/test/java/org/apache/seatunnel/engine/e2e/PipelineConcurrencyIT.java:43-48,61-68,82-87,100-104`
- Why this matters: all four tests pass as long as the expected log text appears. They would still pass if serial / limited execution were violated but the logging still said `[Serial Mode]` or `[Limited Concurrency=2]`.
- Risk: false-green coverage on a scheduler feature. This is not a classic flaky `sleep()` problem, but it is still a weak signal and a real regression hole.
- Better fix: add behavior-level assertions, for example using timestamps or explicit task markers to prove that no more than `N` pipelines are running concurrently.
- Severity: Medium
- Related review history: this is related to @dybyte's earlier request for stronger tests.

# Test stability note
- Rating: `Risk present`
- Basis: `PipelineConcurrencyIT.java:43-48,61-68,82-87,100-104` only checks log strings, so the wait/assertion signal is too weak for the behavior this PR is changing.
- This is not the highest-flakiness pattern, but it is still a test-quality risk because the suite can go green while the scheduler logic is wrong.

# CI note
- The branch is currently `CONFLICTING`.
- The latest `Build` check is still failing, so even aside from the source-level blockers above, the branch is not in a merge-ready state yet.

# Merge conclusion
### Conclusion: can merge after fixes

1. Blocking items
- Issue 1: fix the double increment in `finishedPipelineNum`.
- Issue 2: make the `JobConfig` serialization change backward compatible.
- Issue 3: make the restore / takeover path honor the configured concurrency limit.
- Please also rebase the branch cleanly and get the current `Build` green.

2. Suggested non-blocking follow-up
- Issue 4: strengthen the tests so they verify actual concurrency behavior rather than only log text.
- I would also recommend adding clearer comments around the new scheduler state fields, since this is core engine lifecycle logic.

Overall, I like the goal of making pipeline concurrency configurable, but the current revision still has correctness gaps on the main engine path. Happy to re-review once you push an update.
